### PR TITLE
enhancement(pipelines transform): reduce the number of useless components

### PR DESCRIPTION
**Pull request by jdrouet**
This PR changes the way the `expand` method works by returning the "inner topology" of a component.

Because the way the `expand` was made, it was not possible to control the inputs of each transform that will be expanded. With this inner topology, it's possible to build custom topologies and fine tune the inputs for each inner transform.

From the following configuration

```toml
[sources.random]
type = "random-metrics-and-logs"

[transforms.process]
type = "pipelines"
inputs = ["random"]

[transforms.process.logs]
order = ["foo", "bar"]

[transforms.process.logs.pipelines.foo]
name = "foo"
filter.type = "vrl"
filter.source = """contains(.message, "hello", case_sensitive: false) ?? false"""

[[transforms.process.logs.pipelines.foo.transforms]]
type = "remap"
source = ". |= parse_syslog!(.message)"

[[transforms.process.logs.pipelines.foo.transforms]]
type = "remap"
source = ".went_through_foo = true"

[transforms.process.logs.pipelines.bar]
name = "bar"

[[transforms.process.logs.pipelines.bar.transforms]]
type = "remap"
source = ".went_through_bar = true"

[sinks.output]
type = "console"
inputs = ["process"]
encoding.codec = "json"
target = "stdout"
```

With the previous `expand` method, after each pipeline, there is a `Noop` transform to join the outputs and to be injected in the following pipeline.
In the new `expand` method, considering we can control the input of each created transform, we can specify that the first transform of the next pipeline will take the output of the previous one plus the one that didn't match the previous filter.
On top of this, with the previous `expand` method, we needed, at the end of the pipelines transform to have another `Noop` for the same reason. Now we can just return how the pipelines transform is expanded and by what `process` will be replaced.

### Todo

- [x] add some tests around inputs in pipelines
- [x] add some tests with expanding components in pipelines (like a compound in a pipeline)
- [x] fix unit_test `output_from` getting broken after output
- [ ] more tests

### CODE REVIEW

**Comment by jdrouet** [comment]
Running it locally gave the following results

# Soak Test Results
Baseline: f235412a0b9e76ad1b0cdec863e1b4437dca4f40
Comparison: cc09895a66d71343a9fb5c42cfba1d1ac11a5468
Total Vector CPUs: 4

<details>
<summary>Explanation</summary>
<p>
A soak test is an integrated performance test for vector in a repeatable rig,
with varying configuration for vector.  What follows is a statistical summary of
a brief vector run for each configuration across SHAs given above.  The goal of
these tests are to determine, quickly, if vector performance is changed and to
what degree by a pull request. Where appropriate units are scaled per-core.
</p>

<p>
The table below, if present, lists those experiments that have experienced a
statistically significant change in their throughput performance between
baseline and comparision SHAs, with 95.0% confidence OR
have been detected as newly erratic. Negative values mean that baseline is
faster, positive comparison. Results that do not exhibit more than a
±5% change in mean throughput are discarded. An
experiment is erratic if its coefficient of variation is greater
than 0.1. The abbreviated table will be
omitted if no interesting changes are observed.
</p>
</details>

Changes in throughput with confidence ≥ 95.00% and absolute Δ mean >= ±5%:

| experiment                    | Δ mean   |   Δ mean % | confidence   |
|-------------------------------|----------|------------|--------------|
| http_datadog_filter_blackhole | 5.2MiB   |     113.95 | 100.00%      |


<details>
<summary>Fine details of change detection per experiment.</summary>

| experiment                    | Δ mean   |   Δ mean % | confidence   | baseline mean   | baseline stdev   | baseline stderr   |   baseline outlier % |   baseline CoV | comparison mean   | comparison stdev   | comparison stderr   |   comparison outlier % |   comparison CoV | erratic   | declared erratic   |
|-------------------------------|----------|------------|--------------|-----------------|------------------|-------------------|----------------------|----------------|-------------------|--------------------|---------------------|------------------------|------------------|-----------|--------------------|
| http_datadog_filter_blackhole | 5.2MiB   |     113.95 | 100.00%      | 4.57MiB         | 581.36KiB        | 23.83KiB          |                    0 |       0.124256 | 9.77MiB           | 648.73KiB          | 26.62KiB            |                      0 |        0.0648061 | True      | False              |
</details>
<details>
<summary>Fine details of each soak run.</summary>

| (experiment, variant, run_id)                                                           |   total samples | mean    | std       | min     | median   | p90      | p95      | p99      | max      |   skewness |
|-----------------------------------------------------------------------------------------|-----------------|---------|-----------|---------|----------|----------|----------|----------|----------|------------|
| ('http_datadog_filter_blackhole', 'comparison', 'bbb41b64-3635-4b72-b195-128820c6c9b7') |             198 | 9.81MiB | 641.77KiB | 8.03MiB | 9.83MiB  | 10.57MiB | 10.74MiB | 10.91MiB | 11.62MiB | -0.255751  |
| ('http_datadog_filter_blackhole', 'comparison', '0a0cef27-2c1c-403f-9151-028b8d517dba') |             198 | 9.75MiB | 660.22KiB | 8.06MiB | 9.79MiB  | 10.52MiB | 10.81MiB | 11.1MiB  | 11.3MiB  | -0.213609  |
| ('http_datadog_filter_blackhole', 'comparison', 'e6be49ac-6b50-444c-bd64-312227300c4f') |             198 | 9.74MiB | 644.81KiB | 8.05MiB | 9.77MiB  | 10.49MiB | 10.75MiB | 11.1MiB  | 11.3MiB  | -0.195874  |
| ('http_datadog_filter_blackhole', 'baseline', '58f07856-a616-4f79-961d-79a71ccc6f2f')   |             198 | 4.59MiB | 576.51KiB | 3.25MiB | 4.57MiB  | 5.41MiB  | 5.47MiB  | 5.65MiB  | 5.77MiB  | -0.17964   |
| ('http_datadog_filter_blackhole', 'baseline', 'e12c9da3-7ecf-41cc-b059-74a27063da99')   |             199 | 4.57MiB | 565.86KiB | 3.27MiB | 4.56MiB  | 5.36MiB  | 5.41MiB  | 5.7MiB   | 6.02MiB  | -0.0886541 |
| ('http_datadog_filter_blackhole', 'baseline', '823056d4-700a-4bab-a83e-fc9f79b87aa2')   |             198 | 4.53MiB | 602.7KiB  | 2.59MiB | 4.57MiB  | 5.29MiB  | 5.4MiB   | 5.59MiB  | 5.71MiB  | -0.464822  |
</details>

**Comment by blt** [comment]
This is an _excellent_ result. Well done @jdrouet.

File: src/config/mod.rs [hunk 1]
```diff
@@ -462,14 +488,55 @@ pub struct TestInput {
     pub metric: Option<Metric>,
 }
 
+#[derive(Deserialize, Serialize, Debug, Clone, PartialEq, Eq, Hash)]
+#[serde(untagged)]
+pub enum OneOrMany<T> {
```
Comment (jdrouet) [inline src/config/mod.rs:493]: we could probably move this to another place to use it in another place, but where?

File: src/transforms/pipelines/mod.rs [hunk 2]
```diff
@@ -233,26 +268,52 @@ mod tests {
...
-                "foo.logs.pipelines.foo.truthy.transforms.1",
-                "foo.logs.pipelines.foo.truthy.transforms",
-                "foo.logs.pipelines.foo.truthy",
-                "foo.logs.pipelines.foo.falsy",
-                "foo.logs.pipelines.foo",
-                "foo.logs.pipelines.bar.0",
-                "foo.logs.pipelines.bar",
-                "foo.logs.pipelines",
-                "foo.logs",
-                "foo"
```
Comment (lukesteensen) [inline src/transforms/pipelines/mod.rs:254]: Very nice! The expansion process is a bit complicated and it's difficult to see exactly how big the reduction in components is, but this looks great.

**Review by lukesteensen — COMMENTED** [review]
Awesome work! The results here are super promising.

My main implementation question is around the new `expansions` stuff. We're currently passing it around a lot and it seems to add a somewhat complex second lookup layer in a few places. It's not a blocker, but it would be really nice to try to simplify that somehow, whether it's merging those into the existing maps to provide a single lookup, or encapsulating the new lookup logic in a single method on the config itself (or something like that).

File: src/config/mod.rs (hunk 1 above)
Comment (jdrouet) [reply src/config/mod.rs:493]: The advantage is that we don't have a breaking change regarding the unit tests having an `etract_from` becoming a `Vec`.

**Comment by jdrouet** [comment]
>What's the advantage of this type over just using a `Vec`?

@lukesteensen in the configuration, for now, the `extract_from` is not a `Vec` so moving it to a `Vec` would be a breaking change. This way, it is not.

File: src/config/mod.rs [hunk 3]
```diff
@@ -368,17 +369,35 @@ impl TestDefinition<String> {
...
+                            }
+                        }
+                    } else {
+                        errors.push(format!(
+                            r#"Invalid extract_from target in test '{}': '{}' does not exist"#,
+                            name, from
+                        ));
+                    }
+                }
+                if outputs.is_empty() {
```
[the remark this answers is not in the document]
Comment (jdrouet) [reply src/config/mod.rs]: done :+1:

File: src/config/compiler.rs [hunk 4]
```diff
@@ -6,6 +6,19 @@ use super::{
     builder::ConfigBuilder, graph::Graph, schema, validation, ComponentKey, Config, OutputId,
 };
 
+pub(crate) fn to_string_expansions(
```
[the remark this answers is not in the document]
Comment (jdrouet) [reply src/config/compiler.rs:11]: done 👍

**Review by lukesteensen — APPROVED** [review]
Awesome work! 🚀

**Comment by jdrouet** [comment]
🎉

**Comment by blt** [comment]
Yeeesss very nice work here @jdrouet!

**Pull request merged**